components/TextFields/src/ContainedInputView/MDCBaseTextFieldLayout.m

```diff
@@ -0,0 +1,120 @@
+                         leftViewMode:(UITextFieldViewMode)leftViewMode
+                            rightView:(UIView *)rightView
+                        rightViewMode:(UITextFieldViewMode)rightViewMode
+                                isRTL:(BOOL)isRTL
+                            isEditing:(BOOL)isEditing {
+  self = [super init];
+  if (self) {
+    [self calculateLayoutWithTextFieldSize:textFieldSize
+                                  leftView:leftView
+                              leftViewMode:leftViewMode
+                                 rightView:rightView
+                             rightViewMode:rightViewMode
+                                     isRTL:isRTL
+                                 isEditing:isEditing];
+    return self;
+  }
+  return nil;
+}
+
+#pragma mark Layout Calculation
+
+- (void)calculateLayoutWithTextFieldSize:(CGSize)textFieldSize
+                                leftView:(UIView *)leftView
+                            leftViewMode:(UITextFieldViewMode)leftViewMode
+                               rightView:(UIView *)rightView
+                           rightViewMode:(UITextFieldViewMode)rightViewMode
+                                   isRTL:(BOOL)isRTL
+                               isEditing:(BOOL)isEditing {
+  BOOL shouldAttemptToDisplayLeftView = [self shouldAttemptToDisplaySideView:leftView
+                                                                    viewMode:leftViewMode
+                                                                   isEditing:isEditing];
+  BOOL shouldAttemptToDisplayRightView = [self shouldAttemptToDisplaySideView:rightView
+                                                                     viewMode:rightViewMode
+                                                                    isEditing:isEditing];
+  CGFloat leftViewWidth = CGRectGetWidth(leftView.frame);
+  CGFloat leftViewHeight = CGRectGetHeight(leftView.frame);
+  CGFloat leftViewMinX = 0;
+  CGFloat leftViewMinY = 0;
+  if (shouldAttemptToDisplayLeftView) {
+    leftViewMinX = kHorizontalPadding;
+    leftViewMinY = kTemporaryVerticalPadding;
+  }
+
+  CGFloat rightViewWidth = CGRectGetWidth(rightView.frame);
+  CGFloat rightViewHeight = CGRectGetHeight(rightView.frame);
+  CGFloat rightViewMinX = 0;
+  CGFloat rightViewMinY = 0;
+  if (shouldAttemptToDisplayRightView) {
+    CGFloat rightViewMaxX = textFieldSize.width - kHorizontalPadding;
+    rightViewMinX = rightViewMaxX - rightViewWidth;
+    rightViewMinY = kTemporaryVerticalPadding;
+  }
+
+  CGRect leftViewFrame = CGRectMake(leftViewMinX, leftViewMinY, leftViewWidth, leftViewHeight);
+  CGRect rightViewFrame = CGRectMake(rightViewMinX, rightViewMinY, rightViewWidth, rightViewHeight);
+
+  self.leftViewFrame = leftViewFrame;
+  self.rightViewFrame = rightViewFrame;
+  self.leftViewHidden = !shouldAttemptToDisplayLeftView;
+  self.rightViewHidden = !shouldAttemptToDisplayRightView;
+}
+
+- (BOOL)shouldAttemptToDisplaySideView:(UIView *)subview
```

Comment:
The name of this API implies that there might be situations in which the side view is hidden if we return YES here, but the current code looks like the result directly maps to the `hidden` property. Would it be more accurate to say `showsSideView:(UIView *)sideView ...`?

Reply:
Its original intent was more in line with what the name suggests... I was thinking of having the views not render if there wasn't enough space. I never decided against it, I just never actually built it. I'll change the name here and change it back if/when that behavior actually exists.